process_report/process_report.py

```
@@ -95,6 +99,12 @@ def get_iso8601_time():
     return datetime.datetime.now().strftime("%Y%m%dT%H%M%SZ")
 
 
+def validate_required_env_vars():
```

This function appears to be untested.

---

I added a unit test in `test_util.py`, though given your approval, should I understand that this suggestion was optional?

Also, if you don't mind me rambling on, when does a piece of code become complex enough to warrant a test? I thought this function was too simple to warrant a test. Also, within this repo and a few other repos I worked on, not every function has a test case.

---

In an interpreted, dynamically typed language like Python, if your code doesn't execute during your tests then it could be hiding errors that will crop up at runtime. Consider what would happen if `validate_required_env_vars` looked like this:

```
def validate_required_env_vars():
    for required_env_var in REQUIRED_ENV_VAR:
        if required_env_var not in os.environ:
            sys.exit(f"Required environment variable {required_env_var} is not set")
```

There's a typo there -- `REQUIRED_ENV_VARS` is mis-spelled. As it stands, all your tests pass despite this error. That allows it to sit there undetected until somebody actually runs the code.

The goal of unit tests is to prevent this situation.

You don't necessarily need a test for every individual function, but you do need to ensure that as much code is possible is run as part of your tests. This is where tools like `coverage` come in very handy; if we run the tests using `pytest` and the `coverage` plugin...

```
pytest --cov=process_report --cov-report=html
```

...then we get a nice HTML report that shows exactly what lines of code were covered by tests and what lines were not.